### PR TITLE
feat (audience match types): Update condition evaluator for new audience match types.

### PR DESCRIPTION
**Pull request by yasirfolio3**

### CODE REVIEW

**Review by aliabbasrizvi — COMMENTED** [review]
Changes mostly look good to me. Would leave to @thomaszurkan-optimizely to take a closer look.

**Review by thomaszurkan-optimizely — CHANGES_REQUESTED** [review]
Why are there no unit tests?

File: OptimizelySDKCore/OptimizelySDKCoreTests/TestData/test_data_25_experiments.json [hunk 1]
```diff
@@ -1 +1 @@
...
... "6379061102", "key": "testEventWithMultipleGroupedExperiments"}, {"experimentIds": ["6362042330"], "id": "6385201698", "key": "testEventWithExperimentNotRunning"}, {"experimentIds": ["6361931183"], "id": "6385551103", "key": "testEventWithAudiences"}, {"experimentIds": ["6371041921", "6382251626", "6368671885", "6361743021", "6386460951", "6377131549", "6365780767", "6369512098", "6367473060", "6366023085", "6361931183", "6361100609", "6367902537", "6375231186", "6349682899", "6362042330", "6344617435", "6386411740", "6350472041", "6416416234", "6451651052", "6367842673", "6383770101", "6357622647", "6352512126"], "id": "6386470923", "key": "testEventWithMultipleExperiments"}, {"experimentIds": ["6361931183", "6416416234", "6367473060"], "id": "6386460946", "key": "Total Revenue"}], "variables": [{"id": "6384881128", "key": "someString", "type": "string", "defaultValue": "defaultStringValue"}, {"id": "6384881125", "key": "someBoolean", "type": "bool", "defaultValue": "False"}, {"id": "63234881124", "key": "someInteger", "type": "integer", "defaultValue": "1"}, {"id": "6684886128", "key": "someFloat", "type": "float", "defaultValue": "0.5"}], "revision": "92", "anonymizeIP": false}
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCoreTests/TestData/test_data_25_experiments.json:1]: All "type": "custom_dimension" have been converted to "type": "custom_attribute".

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYBaseCondition.m [hunk 2]
```diff
@@ -27,23 +36,166 @@ + (BOOL) isBaseConditionJSON:(NSData *)jsonData {
...
+            // null for now! We plan on updating the SDKs later in the quarter to return true or false as needed
+            return NULL;
+        }
+        DEFAULT {
+            return NULL;
+        }
+    }
+}
+
+-(nullable NSNumber *)evaluateMatchTypeExact:(NSDictionary<NSString *, NSObject *> *)attributes{
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYBaseCondition.m:49]: https://github.com/optimizely/objective-c-sdk/pull/336/files/ee42573da9a10454d12504f33a5b7eebfac054ed#diff-aa4be2fef28e62bdd60ccbf5094615e6R101
1. Property success defaults to null. 
2. Secondly isEqual is only called if both the value and userAttribute are of similar acceptable types i.e Numeric, Bool, String, Null

Testcase for condition mismatch can be found here:
https://github.com/optimizely/objective-c-sdk/pull/336/files#diff-b36e2d0d8afb118bc8ee016acb8a85cfR166

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYBaseCondition.m [hunk 3]
```diff
@@ -27,23 +36,166 @@ + (BOOL) isBaseConditionJSON:(NSData *)jsonData {
...
+            dict[OPTLYDatafileKeysConditionType] != nil) {
             return true;
         }
         return false;
     }
 }
 
-- (BOOL)evaluateConditionsWithAttributes:(NSDictionary<NSString *, NSObject *> *)attributes {
+- (nullable NSNumber *)evaluateConditionsWithAttributes:(NSDictionary<NSString *, NSObject *> *)attributes {
     if (attributes == nil) {
```
[the remark this answers is not in the document]
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYBaseCondition.m]: This is implemented in PR:
https://github.com/optimizely/objective-c-sdk/pull/337

**Comment by yasirfolio3** [comment]
Changes made, please review.

**Comment by mfahadahmed** [comment]
build

**Review by thomaszurkan-optimizely — APPROVED** [review]
LGTM

**Pull request merged**